packages/compiler-cli/src/ngtsc/metadata/src/global.ts

```diff
@@ -0,0 +1,156 @@
+        1: this.meta.imports?.map((m) => this.reader.getPipeMetadata(m)).filter(nonNull) ?? [],
+        2: this.meta.imports?.map((m) => this.reader.getNgModuleMetadata(m)).filter(nonNull) ?? [],
+      };
+    }
+    return this._imports[kind];
+  }
+
+  public exports<K extends MetaKind>(kind: K): Array<MetaByKind<K>> {
+    if (this.meta.kind === MetaKind.NgModule) {
+      if (this._exports === undefined) {
+        this._exports = {
+          0: this.meta.exports.map((m) => this.reader.getDirectiveMetadata(m)).filter(nonNull),
+          1: this.meta.exports.map((m) => this.reader.getPipeMetadata(m)).filter(nonNull),
+          2: this.meta.exports.map((m) => this.reader.getNgModuleMetadata(m)).filter(nonNull),
+        };
+      }
+      return this._exports[kind];
+    }
+    return [];
+  }
+
+  public all(): Array<PipeMeta|DirectiveMeta|NgModuleMeta> {
+    let result: Array<PipeMeta|DirectiveMeta|NgModuleMeta> = [];
+    for (const ie of ['imports', 'exports'] as const) {
+      for (const kind of [MetaKind.Directive, MetaKind.Pipe, MetaKind.NgModule]) {
+        result.push(...this[ie](kind));
+      }
+    }
+    return result;
+  }
+}
+
+/**
+ * A reader of directive, pipe, and module metadata for all types the compiler has ever seen. This
+ * includes types defined in the current compilation unit, as well as all upstream transitive
+ * exports and re-exports.
+ */
+export class GlobalMetadataReader implements MetadataReaderWithIndex, NgModuleIndex {
+  constructor(
+      private metaReader: MetadataReader, private localMetaReader: MetadataReaderWithIndex) {}
+
+  private indexed = false;
+  private indexFor = new Map<ClassDeclaration, ImportExportIndex>();
+  private ngModulesExportingTrait = new Map<ClassDeclaration, Set<NgModuleMeta>>();
+  private metaForTrait?: {[K in MetaKind]: Map<ClassDeclaration, MetaByKind<K>>};
+
+  private ensureIndex() {
+    if (this.indexed) return;
+    this.indexed = true;
+    this.metaForTrait = {0: new Map(), 1: new Map(), 2: new Map()};
+    const localDirectives =
+        this.localMetaReader.getKnown(MetaKind.Directive)
+            .map(m => this.localMetaReader.getDirectiveMetadata(new Reference(m)));
+    const localPipes = this.localMetaReader.getKnown(MetaKind.Pipe)
+                           .map(m => this.localMetaReader.getPipeMetadata(new Reference(m)));
+    const localNgModules =
+        this.localMetaReader.getKnown(MetaKind.NgModule)
+            .map(m => this.localMetaReader.getNgModuleMetadata(new Reference(m)));
+    let stack = [...localDirectives, ...localPipes, ...localNgModules].filter(nonNull);
+    let visited = new Set(stack.map(t => t.ref.node));
+    while (stack.length > 0) {
+      const trait = stack.pop()!;
+      const index =
+          cache(this.indexFor, trait.ref.node, () => new ImportExportIndex(trait, this.metaReader));
+      cache(this.metaForTrait[trait.kind], trait.ref.node, () => trait);
```

Comment:
This seems really strange to me. An explicit assignment would do the same:

```suggestion
      this.metaForTrait[trait.kind].set(trait.ref.node, trait);
```